wagtail/core/query.py

```diff
@@ -82,6 +82,105 @@ def not_ancestor_of(self, other, inclusive=False):
         """
         return self.exclude(self.ancestor_of_q(other, inclusive))
 
+    def first_common_ancestor(self, include_self=False, strict=False):
```

Comment:
Just to chime in on this (because I did some work on this before @coredumperror took over, and will come up again probably): it was necessary to move this method to avoid duplicating code. The method is needed for checking the permissions but in the current codebase it is a method of PageQuerySet - but the collections extend TreeQuerySet.